### ISSUE TITLE
Add a new honeypot on the registration form that spammers won't be able to answer correctly

### ISSUE THREAD

**Issue by jenlampton** [backdrop-ops/backdropcms.org#895]
This is a follow-up to https://github.com/backdrop-ops/backdropcms.org/issues/815 as another possible solution to the spam problem. 

> The purpose was to not only distinguish between people and bots, but also between click-workers who know nothing about Backdrop and real Backdrop-interested people.

**Comment by jenlampton** [comment]
Today I discovered that a new 'Reason' field was added to the registration form, without testing it also caused problems for people with existing user accounts. It was a required field, preventing anyone with an account from making any changes to their profile without entering something into the new field -- in today's case preventing someone from unsubscribing from the security announcements that just went out.

Thought #1: If we add another field, it will either need to not be required, or be required only on the registration form, and not for new users. (I'm setting this up now for the existing `Biography` field, so we can copy that if it works)

Thought #2: Additionally, having a field called `Reason` (and maybe also the description text on the field) made it feel like we were expecting people to justify why they want an account on backdropcms.org. 

Though this might be a good way to deter click-workers, it will also deter a lot of good-natured Backdrop enthusiasts who might feel that they don't have a strong enough justification to join the community. Having that question on the registration form, to be seen by someone before they have even taken the first step towards joining us, felt unfriendly and against our intent for Backdrop to be a welcoming place for everyone.

I wonder if there might be another question we can ask that might serve the same purpose for administrators to tell the difference between click-workers and Backdrop-interested people. 

My first idea would be something like "What is your favorite Backdrop feature, and why?" that's something that might excite Backdrop-interested people instead of scaring them away, and might also be hard for click-workers to answer.

Thought #3: Another thing to keep in mind is the longer the registration form gets, the fewer real people will register. This decrease in views to conversions increases if any of the fields on the form are required. Because all the fields on our registration form are required, we should strive to keep it as short as possible. 

Once we get this new field onto the registration form, could we remove the Bio field?

**Comment by klonos** [comment]
> ...the longer the registration form gets, the fewer real people will register. This decrease in views to conversions increases if any of the fields on the form are required. Because all the fields on our registration form are required, we should strive to keep it as short as possible.

How about this:

1. we allow people to register with minimum effort (for the purpose of subscribing to newsletters etc.), but we do not allow them to post any content initially.
2. if people want to add themselves to the supporters, or to the contractors for hire, or to add a bio, then they need to follow a different process, which is more "scrutinous" and needs to go through review/approval.

The above workflow would make things harder only for the group of people that are more likely to spam us.

**Comment by jenlampton** [comment]
> we allow people to register with minimum effort (for the purpose of subscribing to newsletters etc.), but we do not allow them to post any content initially.

Creating a user account is the same thing as creating content, because it creates a page on the site. If they go back and edit their user account, they already have access to all the places they'll want to add links to whatever services they were paid to promote.

**Comment by klonos** [comment]
> ...because it creates a page on the site.

That's what I'm asking us to consider changing.

**Comment by jenlampton** [comment]
No  more user profiles?

**Comment by klonos** [comment]
...the various fields on the user profile that are public-facing will be available only to people with a specific user role, which can be added after review (once they have convinced us what their intentions are). We can ask things like "what you like best about Backdrop" as part of the workflow that assigns that elevated role to people.

**Comment by jenlampton** [comment]
This will make things harder for everyone.  Yes, spammers, but also real people who now need to jump through hoops, and admins who now need to monitor the hoops. 

I think we should try to limit our "spam prevention" to things that won't impact everyone else negatively.

**Comment by jenlampton** [comment]
Also -- I created this issue to address the single new field we're trying to add to the registration form :) 

Can you add your ideas about limited access in another issue, or in https://github.com/backdrop-ops/backdropcms.org/issues/815?

**Comment by bugfolder** [comment]
I like the direction that @jenlampton is suggestion for the question. How about this wording (slight tweak):

* What is your favorite feature of Backdrop, or what aspect of Backdrop are you interested in?

And again, the goal of this question is something that a human administrator can use to judge more accurately (than the current "Biography" field) whether an account is being created by a spammer/click worker.

> How about this: ... we allow people to register with minimum effort (for the purpose of subscribing to newsletters etc.), but we do not allow them to post any content initially.

I would really like our defenses to not just prevent spammers from creating pages, but to strive to prevent them from having user accounts at all, if at all possible. The idea being "defense in depth". Even though now, the only thing spammers can create are profile pages, as the site evolves, they might eventually be able to create more. I do totally agree with the point above that we should not do something at the cost of friendliness.

> Once we get this new field onto the registration form, could we remove the Bio field? IIRC we only included it there because we needed a long-form field for Akismet to analyze, but maybe we can have it check this field now instead (or, in addition).

Yes, the bio field was moved to the user reg form just for human analysis; this question will fully serve that role.

**Comment by jenlampton** [comment]
> What is your favorite feature of Backdrop, or what aspect of Backdrop are you interested in?

I love it!

I was also thinking that we should keep it a plain text field instead of filtered, so that spammers won't think adding links or images there will get them anywhere.

**Comment by bugfolder** [comment]
Suggested label for the new field: "Welcome question".

**Comment by bugfolder** [comment]
An update on this, because I've run into a problem. I'm currently working on a local dev copy of b.org. I created the new field as "Text (long)", plain text, and called it just "Welcome" (looked better on the user registration page), but the help text did not show up on the user registration page.

Switching around themes established that the Borg theme drops the help text in this case (and the Backdrop CMS theme inherits the behavior). Using Basis as the default theme, the help text shows up correctly.

So now looking into `*_form_user_login_alter()` and templates to see why it's being left out.

**Comment by bugfolder** [comment]
OK, it's given `display: none;` in file `components.css`:

```
#user-register-form .form-item div.description {
  display: none;
}
```

This wipes out _all_ of the description text on the user registration form, including the help text for the username and email address. So I ask: why? Seems like having help text would be friendlier than not.

**Comment by jenlampton** [comment]
> Suggested label for the new field: "Welcome question".

Wait, I thought we just decided to use "What is your favorite feature of Backdrop, or what aspect of Backdrop are you most interested in?"

> but the help text did not show up on the user registration page.

What are you using for help text?  I don't think we need to tell anyone that this is a question we will use for spam prevention. I think the defeats the purpose...

> OK, it's given `display: none`; in file `components.css`:

This is bad :(  We should open an issue in the borg queue to fix that!

**Comment by jenlampton** [comment]
I've merged the PR and deployed the change to b.org. Thank you @bugfolder for all your work on this!
